### PR TITLE
gdal-utils - parsing refactor

### PR DESCRIPTION
**Pull request by idanmiara**
I've refactored most of `gdal-utils` scripts with the use of an enhanced argparser.
I'm panting the bases for generating the `.rst` files from the `.py` sources to improve the documentation - see `gdal_calc.py` for example for full details.

### CODE REVIEW

**Comment by idanmiara** [comment]
@rouault I'm planning to finish very soon refactoring the remaining of the scripts in the same manner. 
Could we still make it in time for 3.3 ?

**Comment by rouault** [comment]
> Could we still make it in time for 3.3 ?

I intend to issue the release candidate tomorrow. I don't perceive that refactoring work as something essential to complete before 3.3, or am I missing something ?

**Comment by idanmiara** [comment]
> > Could we still make it in time for 3.3 ?
> 
> I intend to issue the release candidate tomorrow. I don't perceive that refactoring work as something essential to complete before 3.3, or am I missing something ?

Nothing essential. I just think it's a neat so I hoped it wouldn't need to be pushed back to 3.4, if it makes sence at this stage. 
No breaking changes. 
I hope I can finish the rest by tommorow.

**Comment by rouault** [comment]
> I hope I can finish the rest by tommorow.

I'd prefer avoiding adding last minute non-trivial changes.

**Comment by idanmiara** [comment]
> > I hope I can finish the rest by tommorow.
> 
> I'd prefer avoiding adding last minute non-trivial changes.

Sure, no biggie.
I'm guessing that by last minute you are also referring to this PR, so in that case I'll break it into a small bugfix PR and the main refactoring of this PR?

**Comment by rouault** [comment]
> I'll break it into a small bugfix PR and the main refactoring of this PR?

yes please

**Comment by idanmiara** [comment]
> > I'll break it into a small bugfix PR and the main refactoring of this PR?
> 
> yes please

I'll rebase this PR after https://github.com/OSGeo/gdal/pull/3742 is merged.

**Comment by idanmiara** [comment]
@rouault this PR is now ready for review/merge. 
I guess you'll be able to merge it after you tag 3.3.0 RC1?

File: autotest/pymod/test_py_scripts.py [hunk 1]
```diff
@@ -67,8 +68,19 @@ def get_py_script(script_name):
...
-    return run_py_script_as_external_script(script_path, script_name, concatenated_argv)
+def run_py_script(script_path: str, script_name: str, concatenated_argv: str,
+                  run_as_script: bool = True, run_as_module: bool = False):
+    result = None
+    if run_as_module:
+        try:
+            module = importlib.import_module('osgeo_utils.' + script_name)
+        except ImportError:
+            module = importlib.import_module('osgeo_utils.samples.' + script_name)
+        argv = [module.__file__] + concatenated_argv.split(' ')
```
Comment (rouault) [inline autotest/pymod/test_py_scripts.py:79]: might be fragile if some arguments include a space and are quoted

Comment (idanmiara) [reply autotest/pymod/test_py_scripts.py:79]: Thanks, I've replaced that split with `shlex.split`.

**Comment by idanmiara** [comment]
@rouault Would you advise me to break down this PR into smaller PRs?

**Comment by rouault** [comment]
There are instances of ``driver`` in that PR too that should be driver_name

**Comment by idanmiara** [comment]
> There are instances of `driver` in that PR too that should be driver_name

OK I think I've replaced that in anyplace I could find now :)

**Pull request merged**